packages/ses/src-xs/compartment.js

```diff
@@ -0,0 +1,284 @@
+      }),
+      ...arrayMap(reexports, specifier => ({ exportAllFrom: specifier })),
+      ...arrayMap(exports, name => ({ export: name })),
+    ],
+  };
+};
+
+const adaptModuleSource = source => {
+  if (source.execute) {
+    return adaptVirtualModuleSource(source);
+  }
+  // eslint-disable-next-line no-underscore-dangle
+  if (source.__syncModuleProgram__) {
+    throw new SyntaxError(
+      'XS native compartments do not support precompiled module sources',
+    );
+  }
+  return source;
+};
+
+const adaptModuleDescriptor = (
+  descriptor,
+  specifier,
+  compartment = undefined,
+) => {
+  if (Object(descriptor) !== descriptor) {
+    throw new TypeError('module descriptor must be an object');
+  }
+  if (descriptor.namespace !== undefined) {
+    return descriptor;
+  }
+  if (descriptor.source !== undefined) {
+    return {
+      source: adaptModuleSource(descriptor.source),
+      importMeta: descriptor.importMeta,
+      specifier: descriptor.specifier,
+    };
+  }
+  // Legacy support for record descriptors.
+  if (descriptor.record !== undefined) {
+    if (
+      descriptor.specifier === specifier ||
+      descriptor.specifier === undefined
+    ) {
+      return {
+        source: adaptModuleSource(descriptor.record),
+        specifier,
+        importMeta: descriptor.importMeta,
+      };
+    } else {
+      if (compartment === undefined) {
+        throw new TypeError(
+          'Cannot construct forward reference module descriptor in module map',
+        );
+      }
+      const { descriptors } = weakmapGet(privateFields, compartment);
+      mapSet(descriptors, descriptor.specifier, {
+        compartment,
+        namespace: specifier,
+      });
+      return {
+        source: adaptModuleSource(descriptor.record),
+        specifier: descriptor.specifier,
+        importMeta: descriptor.importMeta,
+      };
+    }
+  }
+  if (descriptor.specifier !== undefined) {
+    return {
+      namespace: descriptor.specifier,
+      compartment: descriptor.compartment,
+    };
+  }
+  // Legacy support for a source in the place of a descriptor.
+  return { source: adaptModuleSource(descriptor) };
+};
+
+export const adaptCompartmentConstructors = (
+  NativeCompartment,
+  ShimCompartment,
+  maybeHarden,
+) => {
+  function Compartment(...args) {
+    const options = compartmentOptions(...args);
+
+    const {
+      name = undefined,
+      globals = {},
+      transforms = [],
+      resolveHook = () => {
+        throw new TypeError('Compartment requires a resolveHook');
+      },
+      loadHook = undefined,
+      loadNowHook = undefined,
+      importHook = loadHook,
+      importNowHook = loadNowHook,
+      moduleMapHook = () => {},
+      __native__: delegateNative = false,
+      __noNamespaceBox__: noNamespaceBox = false,
+    } = options;
+
+    const modules = delegateNative
+      ? fromEntries(
+          arrayMap(
+            entries(options.modules ?? {}),
+            ([specifier, descriptor]) => [
```

Comment:
Do the imports from commons.js mean that this code needs to be robust against a tampered environment? If so, then
```suggestion
            ({ 0: specifier, 1: descriptor }) => [
```